lib/backend/bdb_ro.c

```diff
@@ -0,0 +1,811 @@
+    type = page[25];
+    if (type != 2 && type != 13 && type != 3 && type != 5)
+	return;
+    nent = *(uint16_t *)(page + 20);
+    if (nent > (pagesize - 26) / 2)
+	nent = (pagesize - 26) / 2;
+    for (i = 0; i < nent; i++) {
+	int minoff = 26 + nent * 2;
+	swap16(page + 26 + i * 2);		/* offset */
+	off = *(uint16_t *)(page + 26 + i * 2);
+	if (off < minoff || off >= pagesize)
+	    continue;
+	if (type == 2 || type == 13) {		/* hash */
+	    if (page[off] == 3 && off + 12 <= pagesize)
+	        swap32_2(page + off + 4);	/* page no/length */
+	} else if (type == 3) {			/* btree internal */
+	    if (off + 12 > pagesize)
+	        continue;
+	    swap16(page + off);			/* length */
+	    swap32_2(page + off + 4);		/* page no/num recs */
+	    if (page[off + 2] == 3 && off + 24 <= pagesize)
+		swap32_2(page + off + 16);	/* with overflow page/length */
+	} else if (type == 5) {			/* btree leaf */
+	    if (off + 3 <= pagesize && page[off + 2] == 1)
+		swap16(page + off);		/* length */
+	    else if (off + 12 <= pagesize && page[off + 2] == 3)
+		swap32_2(page + off + 4);	/* overflow page/length */
+	}
+    }
+}
+
+static int bdb_getpage(struct bdb_db *db, unsigned char *page, unsigned int pageno)
+{
+    if (!pageno || pageno > db->lastpage)
+	return -1;
+    if (pread(db->fd, page, db->pagesize, (off_t)pageno * db->pagesize) != db->pagesize) {
+	perror("pread");
+	return -1;
+    }
+    if (db->swapped)
+	bdb_swappage(db, page);
+    if (pageno != *(uint32_t *)(page + 8))
+	return -1;
+    return 0;
+}
+
+static void bdb_close(struct bdb_db *db)
+{
+    if (db->fd >= 0)
+	close(db->fd);
+    free(db);
+}
+
+static struct bdb_db *bdb_open(const char *name)
+{
+    uint32_t meta[512 / 4];
+    int i, fd;
+    struct bdb_db *db;
+
+    fd = open(name, O_RDONLY);
+    if (!fd) {
+	perror(name);
+	return 0;
+    }
+    db = xcalloc(1, sizeof(*db));
+    db->fd = fd;
+    if (pread(fd, meta, 512, 0) != 512) {
+	perror("pread");
+	bdb_close(db);
+	return 0;
+    }
+    if (meta[3] == 0x00061561 || meta[3] == 0x61150600) {
+	db->type = BDB_HASH;
+	db->swapped = meta[3] == 0x61150600;
+    } else if (meta[3] == 0x00053162 || meta[3] == 0x62310500) {
+	db->type = BDB_BTREE;
+	db->swapped = meta[3] == 0x62310500;
+    } else {
+	fprintf(stderr, "%s: not a berkeley db hash/btree database\n", name);
+	bdb_close(db);
+	return 0;
+    }
+    if (db->swapped)
+        bdb_swapmetapage(db, (unsigned char *)meta);
+    db->pagesize = meta[5];
+    db->lastpage = meta[8];
+    if (db->type == BDB_HASH) {
+	if (meta[4] < 8 || meta[4] > 10) {
+	    fprintf(stderr, "unsupported hash version %d\n", meta[4]);
+	    bdb_close(db);
+	    return 0;
+	}
+	db->maxbucket = meta[18];
+	db->highmask = meta[19];
+	db->lowmask = meta[20];
+	for (i = 0; i < 32; i++)
+	    db->spares[i] = meta[24 + i];
+    }
+    if (db->type == BDB_BTREE) {
+	if (meta[4] < 9 || meta[4] > 10) {
+	    fprintf(stderr, "unsupported btree version %d\n", meta[4]);
+	    bdb_close(db);
+	    return 0;
```

Comment:
All these early returns should be NULLs, not zeros as the function returns a pointer.
I'd also prefer the using the "goto err" idiom to minimize the number of exit points from the function.